Improve Go build process

johananl opened the pull request:
This PR improves the build process for the Go version of the operator, both for plain binaries and for Docker images.

Changes:

- Replace dummy Dockerfile with a real one.
- Add a `.dockerignore` file to exclude irrelevant files from Docker builds.
- Add Make targets for building Go binaries.
- Bake operator version into the binary and print it during initialization.

alourie commented on docker/cassandra-operator/Makefile:
```diff
@@ -1,8 +1,15 @@
-	docker build \
+	cd $(rootdir) && docker build \
+		--build-arg version=$(git_revision) \
 		-t $(DOCKER_REPO)/cassandra-operator \
```
so I guess we should remove `/` here after `DOCKER_REPO`

johananl replied:
Thanks, fixed.

For the record, this is necessary following https://github.com/instaclustr/cassandra-operator/pull/207#issuecomment-518395482.

johananl commented:
I guess we should wait for https://github.com/instaclustr/cassandra-operator/pull/207 before merging as things are currently broken due to the inconsistent handling of slashes in image names.

johananl commented:
TODO before merging:

- [ ] Rebase after https://github.com/instaclustr/cassandra-operator/pull/229 is merged.
- [ ] Modify `Dockerfile` as follows:

```diff
diff --git a/docker/cassandra-operator/Dockerfile b/docker/cassandra-operator/Dockerfile
index 1a63c31..fed5dda 100644
--- a/docker/cassandra-operator/Dockerfile
+++ b/docker/cassandra-operator/Dockerfile
@@ -2,18 +2,8 @@ FROM golang AS builder
 
 ARG version
 
-COPY . /go/src/github.com/instaclustr/cassandra-operator
-WORKDIR /go/src/github.com/instaclustr/cassandra-operator
-
-# Install dep
-RUN go get -d -u github.com/golang/dep \
-    && cd $(go env GOPATH)/src/github.com/golang/dep \
-    && DEP_LATEST=$(git describe --abbrev=0 --tags) \
-    && git checkout $DEP_LATEST \
-    && go install -ldflags="-X main.version=$DEP_LATEST" ./cmd/dep
-
-# Install dependencies
-RUN dep ensure -v -vendor-only
+COPY . /code
+WORKDIR /code
 
 # Build binary
 RUN cd cmd/manager \
```

alourie commented:
now that both dependencies were merged, I think we can get this one finished too.

johananl commented:
Ready for merge as far as I'm concerned.

smiklosovic commented:
hey @johananl thanks! Can you please check as well that corresponding deployment yaml does not miss anything? https://github.com/instaclustr/cassandra-operator/blob/superdupertopsecretrewrite/deploy/operator.yaml

johananl commented:
>hey @johananl thanks! Can you please check as well that corresponding deployment yaml does not miss anything? https://github.com/instaclustr/cassandra-operator/blob/superdupertopsecretrewrite/deploy/operator.yaml

This YAML file is generated by the SDK. I don't see a reason to change this file in this PR, unless I'm missing something.

johananl commented:
Added a tiny change. Using `FROM scratch` requires that we prepend `./` to the command we use to start the operator. I've added that to the manifest under `deploy/`.

smiklosovic commented:
@johananl Thanks a lot, I squashed, rebased (something else I pushed yesterday after you rebased) and merged here https://github.com/instaclustr/cassandra-operator/commit/6e04fab448b27decd3f206d947279efe0c27e19a

I am closing this.
